### PR TITLE
Support access the tracer context in spring gateway filter

### PR DESCRIPTION
**Pull request by darknesstm**
### <Feature description>
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [x] Update the documentation to include this new feature.
- [x] Tests(including UT, IT, E2E) are added to verify the new feature.
- [ ] If it's UI related, attach the screenshots below.

- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. https://github.com/apache/skywalking/issues/10862
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking-java/blob/main/CHANGES.md).

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -19,6 +19,7 @@ Release Notes.
 * Fix possible IllegalStateException when using Micrometer.
 * Support Grizzly Work ThreadPool Metric Monitor
 * Fix the gson dependency in the kafka-reporter-plugin.
+* Support access the sky-walking tracer context in spring gateway filter
```
Comment (wu-sheng) [inline CHANGES.md:22]:
```suggestion
* Support access the tracer context in the spring gateway filter
```

**Comment by wu-sheng** [comment]
Hi, this feature should be good. But you instrument a lot of things, could you try to put a doc about why you are instrumenting? And why there are so many things have to be done?

**Comment by darknesstm** [comment]
> Hi, this feature should be good. But you instrument a lot of things, could you try to put a doc about why you are instrumenting? And why there are so many things have to be done?

I add a [Feature] issues linked to this.
The source is mainly based on the relevant source of `apm-toolkit-trace`, with some simple modifications.

**Comment by wu-sheng** [comment]
I am not sure why, but this PR doesn't drive CI running.

**Comment by darknesstm** [comment]
> I am not sure why, but this PR doesn't drive CI running.

I also see the same problem in #541

**Comment by wu-sheng** [comment]
No, it isn't. That PR is waiting for maintainers' approval to run. But your PR doesn't.

Yours
image

That PR
image

**Comment by kezhenxu94** [comment]
You need to resolve conflicts before we can run the CI

**Comment by wu-sheng** [comment]
> You need to resolve conflicts before we can run the CI

There was no conflict, still not working.

**Comment by darknesstm** [comment]
> > You need to resolve conflicts before we can run the CI
> 
> There was no conflict, still not working.

My Github Actions now works normally...

**Comment by wu-sheng** [comment]
Could you rebase your commits, 14 commits for now...

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-webflux-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/webflux/WebFluxSkyWalkingSegmentIDInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    @Override
+    public Object afterMethod(Class clazz, Method method, Object[] allArguments, Class<?>[] parameterTypes,
+                              Object ret) {
+        return ret;
+    }
+
+    @Override
+    public void handleMethodException(Class clazz, Method method, Object[] allArguments, Class<?>[] parameterTypes,
+                                      Throwable t) {
+        LOGGER.error("Failed to getDefault segment Id.", t);
```
Comment (wu-sheng) [inline apm-sniffer/apm-toolkit-activation/apm-toolkit-webflux-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/webflux/WebFluxSkyWalkingSegmentIDInterceptor.java:51]: What is `default` segment ID? Anything not default?

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-webflux-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/webflux/WebFluxSkyWalkingTraceContextActivation.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+import net.bytebuddy.matcher.ElementMatcher;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.StaticMethodsInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.ClassStaticMethodsEnhancePluginDefine;
+import org.apache.skywalking.apm.agent.core.plugin.match.ClassMatch;
+import org.apache.skywalking.apm.agent.core.plugin.match.NameMatch;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+/**
+ */
```
Comment (wu-sheng) [inline apm-sniffer/apm-toolkit-activation/apm-toolkit-webflux-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/webflux/WebFluxSkyWalkingTraceContextActivation.java:31]: Are you trying to write something?

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-webflux-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/webflux/WebFluxSkyWalkingStaticMethodsAroundInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.StaticMethodsAroundInterceptor;
+import org.springframework.web.server.ServerWebExchange;
+import org.springframework.web.server.ServerWebExchangeDecorator;
+import org.springframework.web.server.adapter.DefaultServerWebExchange;
+
+import java.util.Optional;
+
+public abstract class WebFluxSkyWalkingStaticMethodsAroundInterceptor implements StaticMethodsAroundInterceptor {
+
+    protected static EnhancedInstance getInstance(Object o) {
```
Comment (wu-sheng) [inline apm-sniffer/apm-toolkit-activation/apm-toolkit-webflux-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/webflux/WebFluxSkyWalkingStaticMethodsAroundInterceptor.java:32]: As protected, why `static`?

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-webflux-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/webflux/WebFluxSkyWalkingStaticMethodsAroundInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        if (o instanceof DefaultServerWebExchange && o instanceof EnhancedInstance) {
+            instance = (EnhancedInstance) o;
+        } else if (o instanceof ServerWebExchangeDecorator) {
+            ServerWebExchange delegate = ((ServerWebExchangeDecorator) o).getDelegate();
+            return getInstance(delegate);
+        }
+        return instance;
+    }
+
+    protected static ContextSnapshot getContextSnapshot(Object o) {
```
Comment (wu-sheng) [inline apm-sniffer/apm-toolkit-activation/apm-toolkit-webflux-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/webflux/WebFluxSkyWalkingStaticMethodsAroundInterceptor.java:43]: Same here.

File: docs/en/setup/service-agent/java-agent/Application-toolkit-webflux.md [hunk 6]
```diff
@@ -46,6 +52,28 @@
 ...
 ```
 
+### Fetch trace context relative IDs 
+```java
+    @Override
+    public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain){
+        // fetch trace ID
+        String traceId = WebFluxSkyWalkingTraceContext.traceId(exchange);
```
Comment (wu-sheng) [inline docs/en/setup/service-agent/java-agent/Application-toolkit-webflux.md:60]: Please update all APIs you added. This is an official doc, we should provide a complete version.

**Comment by wuwen5** [comment]
BTW , If the feature of https://github.com/apache/skywalking/issues/10862 is implemented with minimal modifications, it seems that only the `putCorrelation` method needs to be implemented, as the current `WebFluxSkyWalkingOperators` + `TraceContext` can support read operations.

e.g.
```
public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {

        ServerHttpRequest buildRequest = exchange.getRequest()
                .mutate()
                .headers(httpHeaders -> WebFluxSkyWalkingOperators.continueTracing(exchange, () -> {
                    httpHeaders.add("x-trace-id", TraceContext.traceId());
                    httpHeaders.add("x-segment-id", TraceContext.segmentId());
                    httpHeaders.add("x-span-id", String.valueOf(TraceContext.spanId()));
                }))
                .build();
        return chain.filter(exchange.mutate().request(buildRequest).build());
    }
```

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-webflux-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/webflux/WebFluxSkyWalkingTraceContextActivation.java (hunk 3 above)
Comment (darknesstm) [reply apm-sniffer/apm-toolkit-activation/apm-toolkit-webflux-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/webflux/WebFluxSkyWalkingTraceContextActivation.java:31]: I will delete this

**Comment by darknesstm** [comment]
> BTW , If the feature of [apache/skywalking#10862](https://github.com/apache/skywalking/issues/10862) is implemented with minimal modifications, it seems that only the `putCorrelation` method needs to be implemented, as the current `WebFluxSkyWalkingOperators` + `TraceContext` can support read operations.
> 
> e.g.
> 
> ```
> public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
> 
>         ServerHttpRequest buildRequest = exchange.getRequest()
>                 .mutate()
>                 .headers(httpHeaders -> WebFluxSkyWalkingOperators.continueTracing(exchange, () -> {
>                     httpHeaders.add("x-trace-id", TraceContext.traceId());
>                     httpHeaders.add("x-segment-id", TraceContext.segmentId());
>                     httpHeaders.add("x-span-id", String.valueOf(TraceContext.spanId()));
>                 }))
>                 .build();
>         return chain.filter(exchange.mutate().request(buildRequest).build());
>     }
> ```

In this context `WebFluxSkyWalkingOperators.continueTracing` will add a meaningless LocalSpan

**Comment by wu-sheng** [comment]
> In this context WebFluxSkyWalkingOperators.continueTracing will add a meaningless LocalSpan

Make sense. You are correct. Keep what you have built. But please update others accordingly.

**Review by wu-sheng — APPROVED** [review]
LGTM

**Pull request merged**